src/uct/tcp/sockcm/sockcm_ep.c

```diff
@@ -14,26 +14,201 @@
+    dest_addr = (struct sockaddr *) &(ep->remote_addr);
+
+    status = ucs_socket_create(dest_addr->sa_family, SOCK_STREAM,
+                               &ep->sock_id_ctx->sock_id);
+    if (status != UCS_OK) {
+        ucs_debug("unable to create client socket for sockcm");
+        goto out_free;
+    }
+
+    status = UCS_OK;
+    goto out;
+
+out_free:
+    ucs_free(ep->sock_id_ctx);
+out:
+    return status;
+}
+
+static void uct_sockcm_ep_event_handler(int fd, void *arg)
+{
+    ucs_debug("not implemented");
+}
+
 static UCS_CLASS_INIT_FUNC(uct_sockcm_ep_t, const uct_ep_params_t *params)
 {
-    uct_sockcm_iface_t *iface = ucs_derived_of(params->iface,
-                                               uct_sockcm_iface_t);
+    const ucs_sock_addr_t *sockaddr = params->sockaddr;
+    uct_sockcm_iface_t    *iface    = NULL;
+    struct sockaddr *param_sockaddr = NULL;
+    char ip_port_str[UCS_SOCKADDR_STRING_LEN];
+    ucs_status_t status;
+    size_t sockaddr_len;
 
+    iface = ucs_derived_of(params->iface, uct_sockcm_iface_t);
     UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &iface->super);
+
+    if (iface->is_server) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    if (!(params->field_mask & UCT_EP_PARAM_FIELD_SOCKADDR)) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    UCT_SOCKCM_CB_FLAGS_CHECK((params->field_mask &
+			       UCT_EP_PARAM_FIELD_SOCKADDR_CB_FLAGS) ?
+			      params->sockaddr_cb_flags : 0);
+
+    self->pack_cb       = (params->field_mask &
+                           UCT_EP_PARAM_FIELD_SOCKADDR_PACK_CB) ?
+                          params->sockaddr_pack_cb : NULL;
+    self->pack_cb_arg   = (params->field_mask &
+                           UCT_EP_PARAM_FIELD_USER_DATA) ?
+                          params->user_data : NULL;
+    self->pack_cb_flags = (params->field_mask &
+                           UCT_EP_PARAM_FIELD_SOCKADDR_CB_FLAGS) ?
+                          params->sockaddr_cb_flags : 0;
+    pthread_mutex_init(&self->ops_mutex, NULL);
+    ucs_queue_head_init(&self->ops);
+
+    param_sockaddr = (struct sockaddr *) sockaddr->addr;
+    if (UCS_OK != ucs_sockaddr_sizeof(param_sockaddr, &sockaddr_len)) {
+       ucs_error("sockcm ep: unknown remote sa_family=%d", sockaddr->addr->sa_family);
+       status = UCS_ERR_IO_ERROR;
+       goto err;
+    }
+
+    memcpy(&self->remote_addr, param_sockaddr, sockaddr_len);
+
+    self->slow_prog_id = UCS_CALLBACKQ_ID_NULL;
+
+    status = uct_sockcm_ep_set_sock_id(self);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    status = ucs_sys_fcntl_modfl(self->sock_id_ctx->sock_id, O_NONBLOCK, 0); 
+    if (status != UCS_OK) {
+        goto sock_err;
+    }
+
+    status = ucs_socket_connect(self->sock_id_ctx->sock_id, param_sockaddr);
+    if (UCS_STATUS_IS_ERR(status)) {
+        ucs_debug("%d: connect fail\n", self->sock_id_ctx->sock_id);
+        self->conn_state = UCT_SOCKCM_EP_CONN_STATE_CLOSED;
+        goto sock_err;
+    } else {
```

Comment:
this else is not needed

Reply:
@brminich The else case is needed later to check the state of the socket connection attempt during as part of the event handler here - 
https://github.com/openucx/ucx/pull/3570/files#diff-aca878e51b7d5e6eaa75cd20a1375eadR124

Is there some other reason why you think the else is not needed?

Reply:
I meant the line with `} else {`, not the whole code block). You have `goto`  on line 112, so you can just do 
```
  if (UCS_STATUS_IS_ERR(status)) {
        ucs_debug("%d: connect fail\n", self->sock_id_ctx->sock_id);
        self->conn_state = UCT_SOCKCM_EP_CONN_STATE_CLOSED;
        goto sock_err;
  }

  ucs_debug("%d: connect pass/pending\n", self->sock_id_ctx->sock_id);
...
```

Reply:
Thanks. Made this change as well.